### PR TITLE
fix: raises on file-like object with .name not a valid path

### PR DESCRIPTION
**Pull request by scanny**
**Summary**
Fixes: #2308

**Additional context**
Through a somewhat deep call-chain, partitioning a file-like object (e.g. io.BytesIO) having its `.name` attribute set to a path not pointing to an actual file on the local filesystem would raise `FileNotFoundError` when the last-modified date was being computed for the document.

This scenario is a legitimate partitioning call, where `file.name` is used downstream to describe the source of, for example, a bytes payload downloaded from the network.

**Fix**
- explicitly check for the existence of a file at the given path before accessing it to get its modified date. Return `None` (already a legitimate return value) when no such file exists.
- Generally clean up the implementations.
- Add unit tests that exercise all cases.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 ### Fixes
 
+* **Partitioning raises on file-like object with `.name` not a local file path.** Partitioning a file using the `file=` argument with a file-like object (e.g. io.BytesIO) having a `.name` attribute with a value not a path to a file present on the local filesystem raises `FileNotFoundError`. This prevents setting the `.name` attribute to give a `.metadata.filename` value to elements formed from for example a file retrieved from a network location via HTTP.
```
Comment (Coniferish) [inline CHANGELOG.md:13]: This is a little hard to follow. If I'm following correctly, previously `FileNotFoundError` was raised when a file-like object with a `.name` attribute (that was not a path) was partitioned using `file=`, right?

Comment (Coniferish) [reply CHANGELOG.md:13]: What you said in the PR description is clearer:
This fix allows a bytes payload to be given a filename in its `metadata` even though such a file never existed on the local filesystem. This fix explicitly checks for the existence of a file at the given path before accessing it to get its modified date. Return None (already a legitimate return value) when no such file exists.

File: test_unstructured/partition/test_common.py [hunk 2]
```diff
@@ -1,3 +1,7 @@
+import datetime as dt
+import io
+import os
+import pathlib
 from dataclasses import dataclass
 from unittest import mock
```
Comment (Coniferish) [inline test_unstructured/partition/test_common.py:7]: Only potential change I can think of is adding an integration test to show we can partition a `io.BytesIO` "file" via `file=`. Approving in advance, though.

**Review by Coniferish — APPROVED** [review]
Approving in advance (comments are nitpicky).

File: CHANGELOG.md (hunk 1 above)
Comment (scanny) [reply CHANGELOG.md:13]: Yeah, the triggering conditions were these:
1. You call partition_{x}() (pretty much any of them).
2. You use the `file` arg to provide the input document.
3. That file value is a file-like object like `io.BytesIO`, not the `io.BufferedReader` you get back from `with open(path) as f`.
4. You assign a name to the file-like object using `file.name = meaningful_name`. You do that so it shows up in the `metadata.filename` field for each element. Maybe you downloaded the bytes of the document over the network so it doesn't have a path on the filesystem, but you still want to indicate what document the elements came from.

Then when the partitioner went to compute the last-modified date for metadata it would try to access the file at `file.name` to get its modified date without checking whether it actually existed on the filesystem or not.

File: test_unstructured/partition/test_common.py (hunk 2 above)
Comment (scanny) [reply test_unstructured/partition/test_common.py:7]: Yeah, good call. I added one for XLSX as a representative example.
